### DISCUSSION TITLE
Windows post-create-command does not pass '\' correctly to external command

### DISCUSSION THREAD

**Discussion by rasusmilch** [rustic-rs/rustic#1110] [Q&A]
I'm trying to get rustic to correctly pass the file to par2 using the repository config file on a Windows 10 machine. 

In the config file I have:
```
[repository.options]
post-create-command = 'C:\\Users\\rsusmilch\\scoop\\shims\\par2j.exe.exe c -rn1 -rr10 %file'
```
However, it looks like rustic is not correctly passing the file to the external program as the output shows:
```
Parchive 2.0 client version 1.3.3.2 by Yutaka Sawada

Base Directory  : "C:\Users\rsusmilch\"
Recovery File   : "C:\Users\rsusmilch\Usersrsusmilchrustic-backup434dataeded50da651889c1de6d6e62f5ad0447d83bf5475ab79ed3e6fd4658f742a9651c.par2"
CPU thread      : 6 / 12
CPU cache limit : 64 KB, 768 KB
CPU extra       : x64 AVX2 CLMUL
Memory usage    : Auto (1615 MB available), Fast SSD

input file is not found, Usersrsusmilchrustic-backup434dataeded50da651889c1de6d6e62f5ad0447d83bf5475ab79ed3e6fd4658f742a[00:05:00] backing up...                  
[WARN] post-create: command was not successful for filename C:\Users\rsusmilch\rustic-backup\434\data\ed\ed50da651889c1de6d6e62f5ad0447d83bf5475ab79ed3e6fd4658f742a9651c, type data, id ed50da651889c1de6d6e62f5ad0447d83bf5475ab79ed3e6
[00:05:01] backing up...                  
```
I'm new to rust programming and so the source code doesn't help me find where the `%file` output is generated so I can't decipher anything sadly. Is there a configuration issue? Bug? Should I just manually crawl the directory structure and handle the files?

Thanks!

**Comment by rasusmilch** [comment]
Turning on trace logging I see:
```
11:59:42 [DEBUG] (1) rustic_backend::local: calling C:\\Users\\rsusmilch\\scoop\\shims\\par2j.exe.exe c -rn1 -rr10 C:\Users\rsusmilch\rustic-backup\434\snapshots\d4b201eeb1c6b6abcf66c41e3c57d23f60bf464895d6f022ac351516dd1ac75a...
11:59:42 [WARN] post-create: command was not successful for filename C:\Users\rsusmilch\rustic-backup\434\snapshots\d4b201eeb1c6b6abcf66c41e3c57d23f60bf464895d6f022ac351516dd1ac75a, type snapshots, id d4b201eeb1c6b6abcf66c41e3c57d23f60bf464895d6f022ac351516dd1ac75a: exit code: 1
```

**Comment by aawsome** [comment]
@rasusmilch Thanks for asking! 
Note that windows is not fully supported and path handling is one of the issues here..

But in your case, I actually wonder why you are having double backslashes `\` - actually a single one should suffice, IMO.

It also seems to me that your command is only able to handle paths with a double backslash which also surprises me (or I don't know what's going on with windows here...). Actually the command is called directly with the correct file name, see https://github.com/rustic-rs/rustic_core/blob/main/crates/backend/src/local.rs#L118

Note that you can also use `%type` and `%id` to self-construct the filename to use in your script.

**Reply by rasusmilch** [reply]
I agree that they should work without escaping the slashes. However, in my backup config toml file I need to escape `\` and spaces or I get the following error:
```
[WARN] ignoring error O:Test: IO error for operation on O:Test: The system cannot find the file specified. (os error 2)
[WARN] ignoring error EngWK260: IO error for operation on EngWK260: The system cannot find the file specified. (os error 2)
```
My config toml has this listed for the backup source:
```
[[backup.sources]]
source = 'O:\Test Eng\WK260'
label = 'WK260'
```
It only works if I escape the path like this:
```
[[backup.sources]]
source = 'O:\\Test\ Eng\\WK260'
label = 'WK260'
```
Notice that the error shows that `O:Test` stripped out the `\` and stopped at the space in the path. Then it started again after the space and stripped the `\` again giving `EngWK260`.

I suspect a similar thing is happening when the program is outputting the filename, since it seems the passed file according to par2 is:
`C:\Users\rsusmilch\Usersrsusmilchrustic-backup434dataeded50da651889c1de6d6e62f5ad0447d83bf5475ab79ed3e6fd4658f742a9651c.par2`.

You can see my directory of the repository is stripped of the drive letter `C:` and `\`.
For reference my repo is `repository = 'C:\Users\rsusmilch\rustic-backup\434'` which DOES work with single `\` and needs no escaping.

Strange...

**Reply by simonsan** [reply]
The first `\` escapes the second `\`. You can also try using a forward slash `/` on windows. Instead of the escaped `\`, it should work as well.

**Reply by aawsome** [reply]
Ah, I got it. `shellword`'s `split` is used to split the the actual command and this handles escaped strings... This is why you need to escape the `\` in your command.
Unfortunately it is used *after*  the variables are substituted. And in the variable substitutions only single `\` are inserted as the paths are taken literally.

Maybe we should apply the substitutions after splitting with `shellwords`?

**Reply by fbewivpjsbsby** [reply]
I can run par2 using this command. But is says "The process cannot access the file because it is being used by another process". I've added ```rustic.exe``` and ```par2.exe``` to Windows Defender exclusion.
```
'cmd /c "par2 create -qq -n1 -r5 %file"'
```
```
"par2 create -qq -n1 -r5 \"%file\""
```

### Log
```
[INFO] using config .\rustic.toml
[DEBUG] (1) rustic_backend::local: calling cmd /c "par2 create -qq -n1 -r5 C:\rustic\keys\bd8fbe20e67
7e9ff3ab65bd4a3844eba866d6a5714741a0b087abd192b3813ad"...

Could not open "C:\rustic\keys\bd8fbe20e677e9ff3ab65bd4a3844eba866d6a5714741a0b087abd192b3813ad": The process cannot access the file because it is being used by another process.

[WARN] post-create: command was not successful for filename C:\rustic\keys\bd8fbe20e677e9ff3ab65bd4a3
844eba866d6a5714741a0b087abd192b3813ad, type keys, id bd8fbe20e677e9ff3ab65bd4a3844eba866d6a5714741a0b087abd
192b3813ad: exit code: 6
[INFO] key bd8fbe20 successfully added.
[DEBUG] (1) rustic_backend::local: calling cmd /c "par2 create -qq -n1 -r5 C:\rustic\config"...

Could not open "C:\rustic\config": The process cannot access the file because it is being used by another process.

[WARN] post-create: command was not successful for filename C:\rustic\config, type config, id 1c9ae96
108cb140ae6dd2fba3f20a7407cf474562043b15ff5b8312004961079: exit code: 6
[INFO] repository 71c63964 successfully created.
```

### Config
```
[global]
log-level = "debug" # any of "off", "error", "warn", "info", "debug", "trace"; default: "info"
dry-run = false

[repository]
repository = "C:\\rustic"
password = "test"

[repository.options]
post-create-command = 'cmd /c "par2 create -qq -n1 -r5 %file"'
# post-create-command = "par2 create -qq -n1 -r5 \"%file\""

post-delete-command = "sh -c \"rm -f %file*.par2\""
```
